File: backend/app/services/openai_service.py
"""OpenAI Whisper transcription + GPT-4o Structured Output service."""

from openai import AsyncOpenAI
from pydantic import BaseModel

from app.core.config import settings

client = AsyncOpenAI(api_key=settings.OPENAI_API_KEY)


class ProductInfo(BaseModel):
    brand: str | None = None
    code: str | None = None
    area: str | None = None


class TreatmentExtraction(BaseModel):
    """Structured output schema for extracting treatment info from voice memos."""

    customer_name: str | None = None
    service_type: str | None = None
    products_used: list[ProductInfo] | None = None
    area: str | None = None
    duration_minutes: int | None = None
    satisfaction: str | None = None
    next_visit_recommendation: str | None = None
    summary: str | None = None


async def transcribe_audio(audio_file_path: str) -> str:
    """Transcribe audio file using OpenAI Whisper API."""
    with open(audio_file_path, "rb") as audio_file:
        transcription = await client.audio.transcriptions.create(
            model="gpt-4o-mini-transcribe",
            file=audio_file,
            language="ko",
        )
    return transcription.text


async def extract_treatment_info(transcript: str) -> TreatmentExtraction:
    """
    Extract structured treatment information from a transcript
    using GPT-4o Structured Output.
    """
    completion = await client.beta.chat.completions.parse(
        model="gpt-4o",
        messages=[
            {
                "role": "system",
                "content": (
                    "당신은 한국 미용실 시술 기록 전문 AI 어시스턴트입니다. "
                    "음성 메모 텍스트에서 시술 정보를 정확하게 추출하세요. "
                    "브랜드명과 제품 코드를 정확히 구분하세요. "
                    "예: '로레알 7.1' → brand='로레알', code='7.1'"
                ),
            },
            {
                "role": "user",
                "content": f"다음 음성 메모에서 시술 정보를 추출해주세요:\n\n{transcript}",
            },
        ],
        response_format=TreatmentExtraction,
    )
    return completion.choices[0].message.parsed


async def transcribe_and_extract(audio_file_path: str) -> TreatmentExtraction:
    """Transcribe audio and extract structured treatment info in one step."""
    transcript = await transcribe_audio(audio_file_path)
    extraction = await extract_treatment_info(transcript)
    return extraction

File: frontend/src/lib/services/openai-service.ts
import OpenAI from "openai";
import { zodResponseFormat } from "openai/helpers/zod";
import { z } from "zod";

function getOpenAI() {
  return new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
}

const ProductInfo = z.object({
  brand: z.string().nullable().optional(),
  code: z.string().nullable().optional(),
  area: z.string().nullable().optional(),
});

const TreatmentExtraction = z.object({
  customer_name: z.string().nullable().optional(),
  service_type: z.string().nullable().optional(),
  products_used: z.array(ProductInfo).nullable().optional(),
  area: z.string().nullable().optional(),
  duration_minutes: z.number().nullable().optional(),
  satisfaction: z.string().nullable().optional(),
  next_visit_recommendation: z.string().nullable().optional(),
  summary: z.string().nullable().optional(),
});

export type TreatmentExtractionResult = z.infer<typeof TreatmentExtraction>;

export async function transcribeAudio(audioBuffer: Buffer, filename: string): Promise<string> {
  const file = new File([new Uint8Array(audioBuffer)], filename, { type: "audio/webm" });
  console.log(`[transcribeAudio] 파일 생성: ${filename}, size: ${audioBuffer.length} bytes`);

  const transcription = await getOpenAI().audio.transcriptions.create({
    model: "whisper-1",
    file,
    language: "ko",
  });

  console.log(`[transcribeAudio] 변환 결과: "${transcription.text}"`);

  if (!transcription.text || transcription.text.trim().length === 0) {
    throw new Error("음성이 인식되지 않았습니다. 더 크게 말하거나 더 길게 녹음해주세요.");
  }

  return transcription.text;
}

export async function extractTreatmentInfo(transcript: string): Promise<TreatmentExtractionResult> {
  const completion = await getOpenAI().chat.completions.create({
    model: "gpt-4o",
    messages: [
      {
        role: "system",
        content:
          "당신은 한국 미용실 시술 기록 전문 AI 어시스턴트입니다. " +
          "음성 메모 텍스트에서 시술 정보를 정확하게 추출하세요. " +
          "브랜드명과 제품 코드를 정확히 구분하세요. " +
          "예: '로레알 7.1' → brand='로레알', code='7.1'" +
          "\n\n반드시 JSON 형식으로 응답하세요.",
      },
      {
        role: "user",
        content: `다음 음성 메모에서 시술 정보를 추출해주세요:\n\n${transcript}`,
      },
    ],
    response_format: zodResponseFormat(TreatmentExtraction, "treatment_extraction"),
  });
  const content = completion.choices[0].message.content;
  return TreatmentExtraction.parse(JSON.parse(content || "{}"));
}

export async function transcribeAndExtract(audioBuffer: Buffer, filename: string): Promise<TreatmentExtractionResult> {
  const transcript = await transcribeAudio(audioBuffer, filename);
  return extractTreatmentInfo(transcript);
}
